src/prime_server.cpp

```diff
@@ -15,6 +21,46 @@ struct interrupt_t : public std::runtime_error {
   }
 };
 constexpr uint32_t INTERRUPT_AGE_CUTOFF = 600; // request age in seconds
+
+struct quiescable final {
+  static const quiescable& get(unsigned int drain_seconds = 0, unsigned int shutdown_seconds = 0) {
+    static quiescable instance(drain_seconds, shutdown_seconds);
+    return instance;
+  }
+  quiescable(unsigned int drain_seconds, unsigned int shutdown_seconds)
+      : draining(false), shutting_down(false) {
+    // if both are unset we disable this functionality
+    if (drain_seconds == 0 && shutdown_seconds == 0)
+      return;
+    // first we block SIGTERM in the main thread (its children will inherit this)
+    sigset_t set;
+    sigemptyset(&set);
+    sigaddset(&set, SIGTERM);
+    auto s = pthread_sigmask(SIG_BLOCK, &set, NULL);
+    if (s != 0) {
+      logging::ERROR("Could not mask SIGTERM, graceful shutdown disabled");
+      return;
+    }
+    // then we make daemon thread just to handle SIGTERM
+    std::thread([set, drain_seconds, shutdown_seconds, this]() {
+      // wait for SIGTERM to trigger
+      int sig = 0;
+      if (sigwait(&set, &sig) == 0) {
+        draining = true;
+        sleep(drain_seconds);
+        shutting_down = true;
+        sleep(shutdown_seconds);
+        exit(0);
+      } else {
+        logging::ERROR("Could not wait for SIGTERM, graceful shutdown disabled");
+      }
+    }).detach();
+  }
+
+  bool draining;
```

Comment:
should these be made an `atomic_bool`s? I guess the downside is 1 extra loop of the main loop of prime server (or could there be more subtle caching issues on a multi-core CPU? maybe I'm overthinking this :smile: ).

Reply:
i considered it but didnt think it was needed because:

1. only the signal handler writes to while other threads are reading it
2. writes to primatives are atomic on modern architectures

Here's a quick explanation of what I'm relying on: https://stackoverflow.com/questions/52424277/c-is-the-assignment-of-a-value-to-a-primitive-data-type-e-g-bool-an-atomic

I'll benchmark it with atomic and see if it matters at all (hopefully not)

Reply:
yeah you're right, there's no danger of partial writes. I was more thinking about different threads on different cores not seeing the writes to the bools because of caching (the cache coherency issue that the SO thread alluded to as well).

Reply:
atomic was fine so i added it. going to merge. ive tested this extensively with the branch i have going in valhalla. preparing this for release

Reply:
Thanks for adding this feature @kevinkreiser :pray: !